### DISCUSSION TITLE
Use codecov to annotate github pull requests?

### DISCUSSION THREAD

**Discussion by boxydog** [getpelican/pelican#3324] [General]
How do people feel about annotating github pull requests with code coverage? See https://docs.codecov.com/v4.6/docs/github-2-getting-a-codecov-account-and-uploading-coverage#github-actions.

We'd have to authorize codecov as a third-party application in github. See https://docs.codecov.com/v4.6/docs/github-2-getting-a-codecov-account-and-uploading-coverage#create-a-codecov-account.

**Comment by boxydog** [comment]
The greatest advantage would be to set coverage standards, either overall for the project, or for patches. See https://docs.codecov.com/v4.6/docs/github-5b-setting-coverage-standards-with-flags

**Comment by justinmayer** [comment]
Most of my experience with Codecov was from several years ago, so I have no idea what (if anything) has changed during that time period. With that caveat aside…

I don't recall having great experiences with Codecov, which is why Rémy and I decided to handle this topic in a different manner for Kagi, a multi-factor authentication system for Django. We decided to aim for 100% test coverage from the start, [adding command-line flags to Pytest that cause test failure if coverage drops below 100%](https://github.com/justinmayer/kagi/blob/e4ffd68812362b1ad0ebb147e702119876cee0ff/tasks.py#L55-L56). Regardless of whether we sought 100% test coverage for Pelican, one could replace the relevant `100` value with the current test coverage, causing a test failure if it drops below that value (i.e., `--cov-fail-under 75`). In addition to local test failure if coverage drops below the specified value, it also fails in a CI context since the GitHub Actions workflow is running [the same test invocation](https://github.com/justinmayer/kagi/blob/e4ffd68812362b1ad0ebb147e702119876cee0ff/.github/workflows/main.yml#L63).

For what it's worth, my friend Hynek also ran into issues with Codecov and came up with a solution that doesn’t rely on third-party services, as he documented in a detailed post called [How to Ditch Codecov for Python Projects](https://hynek.me/articles/ditch-codecov-python/) that presumably describes [how he implemented that solution for `structlog`](https://github.com/hynek/structlog/pull/369).

Just my two cents. I'm not totally opposed to Codecov, but it's probably clear by now that I do like the idea of achieving a similar result without relying on a third-party service.

**Comment by boxydog** [comment]
I also like the thought of no third-party service, as long as it is easy to implement, enforce, and see. I'll read.
